### PR TITLE
dev/core#667 Fix bug where entity_tags are not deleted by 'soft FK'

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fixes a bug where deleting a contact does not delete tags associated with the contact

Before
----------------------------------------
1) Create a contact with a tag
2) Delete the contact (don't delete to trash, full delete)
3) check entity_tag record still exists in civicrm_entity_tag

After
----------------------------------------
entity_tag record is cleaned up

Technical Details
----------------------------------------
This utilises an approach @colemanw & @totten & I discussed about switching to a pre hook rather than ad hoc function calls. It's an initial foray & it would be good to limit it to being called on delete. Other entities could do with this - e.g the participants are deleted & then the participant notes in the current flow - I'm pretty sure the note  won't be deleted due to the participant going first.

Comments
----------------------------------------

### CODE REVIEW

**Comment by seamuslee001** [comment]
This looks fine to me would appreciate @totten just doing a review as well

File: CRM/Core/BAO/EntityTag.php [hunk 1]
```diff
@@ -483,4 +483,37 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
...
+   *
+   * Ideally we will later nuance it so it's only called on delete....
+   *
+   * @param \Civi\Core\Event\PreEvent $event
+   */
+  public static function preDeleteOtherEntity($event) {
+    if ($event->action !== 'delete'
+      // Activity can call the pre hook for delete with no ID - this seems to be isolated to activity....
+      // @todo - what is the correct way to standardise?
+      || ($event->entity === 'Activity' && !$event->id)) {
```
Comment (colemanw) [inline CRM/Core/BAO/EntityTag.php]: Wait... what?
Why would this hook be called with no id? Is that a bug?

File: CRM/Core/BAO/EntityTag.php [hunk 2]
```diff
@@ -483,4 +483,37 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
...
+      $entity = in_array($event->entity, ['Individual', 'Household', 'Organization']) ? 'Contact' : $event->entity;
+      $tableName = CRM_Core_DAO_AllCoreTables::getTableForEntityName($entity);
+      $used_for = CRM_Core_OptionGroup::values('tag_used_for');
+      Civi::$statics[__CLASS__]['tagged_entities'][$event->entity] = !empty($used_for[$tableName]) ? $tableName : FALSE;
+    }
+
+    if (Civi::$statics[__CLASS__]['tagged_entities'][$event->entity]) {
+      CRM_Core_DAO::executeQuery('DELETE FROM civicrm_entity_tag WHERE entity_table = %1 AND entity_id = %2',
+        [1 => [Civi::$statics[__CLASS__]['tagged_entities'][$event->entity], 'String'], 2 => [$event->id, 'Integer']]
+      );
```
Comment (colemanw) [inline CRM/Core/BAO/EntityTag.php]: Firing off an ad-hoc query feels icky somehow. Shouldn't we call pre/post hooks before deleting entityTags? And shouldn't there already be a bao function to delete entityTags? And can we avoid an infinite loop in the process of calling a pre hook from within a pre hook listener?

File: CRM/Core/BAO/EntityTag.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Core/BAO/EntityTag.php]: It feels like it to me - but in activity BAO

```
  public static function deleteActivity(&$params, $moveToTrash = FALSE) {
    // CRM-9137
    if (!empty($params['id']) && !is_array($params['id'])) {
      CRM_Utils_Hook::pre('delete', 'Activity', $params['id'], $params);
    }
    else {
      CRM_Utils_Hook::pre('delete', 'Activity', NULL, $params);
    }
```

File: CRM/Core/BAO/EntityTag.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Core/BAO/EntityTag.php]: good point about the loop.

My instinct here was that this is kind of a bottom-of-the-food-chain entity  &  the efficiency of a single query makes sense

Comment (eileenmcnaughton) [reply CRM/Core/BAO/EntityTag.php]: Oh - the existing entity-tag functions hurt my head

**Comment by colemanw** [comment]
Hang on, it looks like there was a `civi.dao.preDelete` event added a couple years ago. It's a bit lower-level and might be more appropriate than "hook_pre". https://github.com/civicrm/civicrm-core/blob/5.24/CRM/Core/DAO.php#L596-L609

**Comment by eileenmcnaughton** [comment]
ohhh

**Comment by eileenmcnaughton** [comment]
@colemanw so I feel like to merge this we need to resolve the following

1) should we use that preDelete hook - I didn't see in in your PR that I merged
2) is it OK the for kinda 'trivial' entities to not be individual deletes with hooks?

Performance wise it's a bit mixed - doing a get before any deletes does reduce locking operations but if a contact had hundreds of tags individual deletes would be a bit of a drag.

**Comment by colemanw** [comment]
So for:

1. Yes the preDelete hook gets called further down, inside `$dao->delete()` which is called by the new `deleteRecord` function.
2. I guess it's a judgement call and there may be some grey areas, but since `entityTag` is essentially a join table, this seems like a pretty clear-cut "yes".

I guess it doesn't need to be in this PR, but I'd like to see a dispatcher system that makes it easier for entities to subscribe to create/update/delete events fired by other entities.

**Comment by eileenmcnaughton** [comment]
@colemanw I've updated it to use the other listener

File: CRM/Core/BAO/EntityTag.php [hunk 3]
```diff
@@ -483,4 +483,40 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
     return $options;
   }
 
+  /**
+   * This function deletes entity tags when a related entity is called.
+   *
+   * It is registered as a listener in \Civi\Core\Container::createEventDispatcher
+   *
+   * Ideally we will later nuance it so it's only called on delete....
```
Comment (colemanw) [inline CRM/Core/BAO/EntityTag.php]: That's done :)

Comment (eileenmcnaughton) [reply CRM/Core/BAO/EntityTag.php]: @colemanw good point - I've removed the comment - is this merge-on-pass now?

File: CRM/Core/DAO/AllCoreTables.php [hunk 4]
```diff
@@ -254,6 +254,17 @@ public static function getTableForClass($className) {
...
+  /**
+   * Convert the entity name into a table name.
+   *
+   * @param string $entityBriefName
+   *
+   * @return FALSE|string
+   */
+  public static function getTableForEntityName($entityBriefName) {
+    return self::getTableForClass(self::getFullName($entityBriefName));
+  }
```
Comment (colemanw) [inline CRM/Core/DAO/AllCoreTables.php]: This new function seems unrelated to this PR? Was it for something else?

Comment (eileenmcnaughton) [reply CRM/Core/DAO/AllCoreTables.php]: Hmm - the change in hook means we don't need it for this - because now we have a class name not an  entity name

**Comment by colemanw** [comment]
Looks good!

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Hmm it failed around the same place

`not ok 139 - Error: CRM_Case_XMLProcessor_ProcessTest::testCreateActivityWithDefaultContactByRelationship
not ok 140 - Error: CRM_Case_XMLProcessor_ProcessTest::testCreateActivityWithDefaultContactByRelationshipMissing
not ok 141 - Error: CRM_Case_XMLProcessor_ProcessTest::testCreateActivityWithDefaultContactByRelationshipBidirectional
not ok 142 - Error:

**Comment by eileenmcnaughton** [comment]
Ah it's because I deleted that fn you questioned - but it IS being called
